pkgs/test/named-hierarchy/default.nix

```diff
@@ -0,0 +1,184 @@
+              optionalAttrNameError
+              ++ optionalDirectoryError
+              ++ optionalPackageNixError
+            else
+              [ fileError ];
+        in
+        lib.optional (errors != []) ''
+          - pkgs/by-name/${shard}/${attr}:
+          ${lib.concatStrings errors}''
+      ) entries);
+    in
+    if type == "directory" then
+      optionalShardNameError
+      ++ duplicateErrors
+      ++ packageErrors
+    else
+      [ "- pkgs/by-name/${shard}: Is a file, but should be a directory.\n" ];
+
+  shards = builtins.removeAttrs (builtins.readDir baseDirectory) [ "README.md" ];
+
+  structureErrors = lib.concatLists (lib.mapAttrsToList errorPartsForShard shards);
+
+  structureResult =
+    if structureErrors == [] then
+      builtins.trace "Structural evaluation checks for pkgs/by-name successfull" null
+    else
+      throw ''
+        Structural evaluation checks for pkgs/by-name failed:
+        ${lib.concatStrings structureErrors}'';
+
+  packageFiles = import ../../top-level/named-hierarchy-files.nix;
+
+  overlay = self: super: {
+    callPackage = fn: args:
+      let
+        result = super.callPackage fn args;
+      in
+      if lib.isAttrs result then
+        # If this was the last overlay to be applied, we could just only return the `_callPackageFile`, but that's not the case because stdenv has another overlays on top of user-provided ones
+        # So to not break the stdenv build we need to return the proper result here
+        result // {
+          _callPackage.result = result;
+          _callPackage.file = fn;
+        }
+      else
+        # It's very rare that callPackage doesn't return an attribute set, but it can occur.
+        {
+          _callPackage.result = result;
+          _callPackage.file = fn;
+        };
+  };
+
+  extended = extend overlay;
+
+  semanticErrors = lib.concatLists (lib.mapAttrsToList (name: expectedFile:
+    let
+      value = extended.${name};
+      inherit (value._callPackage) file result;
+      prefix = ''
+        - pkgs.${name} (defined by ${lib.removePrefix (toString ../../.. + "/") (toString expectedFile)})'';
+    in
+    if ! value ? _callPackage then
+      [ "${prefix}: Not defined using callPackage" ]
+    else if ! builtins.isPath file then
+      [ "${prefix}: callPackage'd with a non-path but a ${builtins.typeOf file}" ]
+    else if file != expectedFile then
+      [ "${prefix}: Not of the form callPackage ${toString expectedFile}, but rather callPackage ${toString file}" ]
+    else if ! lib.isDerivation value then
+      [ "${prefix}: Not a derivation" ]
+    else
+      [ ]
+  ) packageFiles);
+
+
+  # Excludes the reference check
+  semanticResult =
+    if semanticErrors == [] then
+      builtins.trace "Semantic evaluation checks for pkgs/by-name successfull" null
+    else
+      throw ''
+        Semantic evaluation checks for pkgs/by-name failed:
+        ${lib.concatStrings semanticErrors}'';
+
```

Comment:
That's a lot of untested code up to this point. How do we know that this tests what it's supposed to test? - Does it really catch mistakes?

I'd recommend to factor these functions out and give them tests.

Also, or alternatively, you could use [testBuildFailure](https://nixos.org/manual/nixpkgs/stable/#tester-testBuildFailure) to test the end result of the derivation below, if you make the `baseDirectory` a parameter, and then feed it some bad inputs that way.

Reply:
Yes there's a TODO in the PR description to "Test the checks", will definitely do this!

Reply:
I now added tests for all kinds of errors that could occur.